### PR TITLE
added file upload routes using multer

### PR DESCRIPTION
**Pull request by Sharif37**
user can upload multiple image of a product . Each product_id mach multiple image in one to many relationship . user can upload **8** images at a time . here i use **multer** middle-ware for upload file .

### CODE REVIEW

**Comment by Sharif37** [comment]
In this commit , i  added pagination and filtration   for all routes . i make it generalize for each routes . using pagination you can set custom limit and fetch portion of data from server each time . using filtration you can filters data based on all possible criteria . it covers all possible searching  done by users .

File: src/controllers/productController.ts [hunk 1]
```diff
@@ -87,8 +101,9 @@ export async function deleteProduct(req: Request, res: Response) {
 // Fetch all product categories
 export async function getAllCategories(req: Request, res: Response) {
```
Comment (tajmaun) [inline src/controllers/productController.ts]: Follower proper name conventions 
req -> categroryRequest
res -> categoryResponse

File: src/controllers/productController.ts [hunk 2]
```diff
@@ -98,9 +113,18 @@ export async function getAllCategories(req: Request, res: Response) {
 export async function getProductsByCategory(req: Request, res: Response) {
```
Comment (tajmaun) [inline src/controllers/productController.ts]: Follower proper name conventions
req -> productRequest
res -> productResponse

File: src/controllers/productController.ts [hunk 3]
```diff
@@ -98,9 +113,18 @@ export async function getAllCategories(req: Request, res: Response) {
...
   try {
     const categoryId = parseInt(req.params.categoryId);
-    const products = await productService.getProductsByCategoryId(categoryId);
```
Comment (tajmaun) [inline src/controllers/productController.ts:101]: Add log here

File: src/helper/addCategoryFilters.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  if (filters.category_name) {
+    query = query.where(
+      sql`LOWER(product_category.category_name)`,
+      "like",
+     `%${filters.category_name.toLowerCase()}%`
+    );
+  }
+
+  if (filters.parent_category_id) {
+    query = query.where("product_category.parent_category_id", "=", filters.parent_category_id);
```
Comment (tajmaun) [inline src/helper/addCategoryFilters.ts]: be specific what sort query it is

File: src/helper/addFiltration.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import { addCategoryFilters } from "./addCategoryFilters";
+
+
+export function addFiltration(
+  table: TableName,
+  query: SelectQueryBuilder<DB, TableName, any>,
+  req: Request,
+) {
+
+  if (table === "product") {
```
Comment (tajmaun) [inline src/helper/addFiltration.ts]: "product" declare it as constant

File: src/helper/addFiltration.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    query = addProductFilters(req, query as any);
+  }
+   else if(table === "product_category"){
```
Comment (tajmaun) [inline src/helper/addFiltration.ts]: "product_category" declare it as constant if intend to use it again somewhere in the codebase

File: src/helper/addProductFilters.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import { Request } from "express";
+import { SelectQueryBuilder , sql} from "kysely";
+import { z } from "zod";
+import { DB } from "../database"; 
+
+export function addProductFilters(
+  req: Request,
```
Comment (tajmaun) [inline src/helper/addProductFilters.ts]: be specific in naming

File: src/helper/addProductFilters.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  query: SelectQueryBuilder<DB, "product", {}>
+) {
+  // Filter by product_id (number)
```
Comment (tajmaun) [inline src/helper/addProductFilters.ts:27]: z.coerce.number().safeParse(req.query.product_id).success
 
do this operation separately and use a variable with proper naming to understand its action.

File: src/helper/addProductFilters.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  if (
+    req.query.product_id &&
+    z.coerce.number().safeParse(req.query.product_id).success
+  ) {
+    query = query.where("product.product_id", "=", z.coerce.number().parse(req.query.product_id));
+  }
+
+  // Filter by product_name (string)
+  if (
```
Comment (tajmaun) [inline src/helper/addProductFilters.ts]: same goes to here.

follow above comment

File: src/helper/paginatedResults.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
+import { Request, Response } from "express";
+import { SelectQueryBuilder } from "kysely";
+import { TableName } from "../database";
+import { DB } from "../database";
+import { SQLError } from "../types/SQLError";
+
+export async function paginatedResults(
+  query: SelectQueryBuilder<DB, TableName, any>,
+  req: Request,
```
Comment (tajmaun) [inline src/helper/paginatedResults.ts:9]: use proper naming

**Review by tajmaun — CHANGES_REQUESTED** [review]

**Review by tajmaun — APPROVED** [review]
Looks good

**Pull request merged**